src/Compilers/Core/Portable/DiagnosticAnalyzer/CachingSemanticModelProvider.cs

```diff
@@ -0,0 +1,40 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System.Collections.Concurrent;
+
+namespace Microsoft.CodeAnalysis.Diagnostics
+{
+    internal sealed class CachingSemanticModelProvider : SemanticModelProvider
+    {
+        private readonly ConcurrentDictionary<SyntaxTree, SemanticModel> _semanticModelsMap;
+
+        public CachingSemanticModelProvider()
+        {
+            _semanticModelsMap = new ConcurrentDictionary<SyntaxTree, SemanticModel>();
```

Comment:
>ConcurrentDictionary [](start = 37, length = 20)

What is the lifetime of this dictionary? It looks like it is going to keep added trees alive for as long as it is alive.

Reply:
The dictionary will stay alive for the lifetime of the provider, but the entries will be removed from the map by the analyzer driver whenever analysis completes for each tree.

Reply:
I have added a doc comment to the type for clarity.